### PR TITLE
sjawhar-legion-92: add controller support for multi-repo issue routing

### LINKED ISSUE #92: Add controller support for multi-repo issue routing
## Problem

The daemon is bound to a single GitHub project board (`LEGION_ID`). Issues tracked on different project boards are invisible to the state machine — `/state/fetch-and-collect` only fetches from one board.

Additionally, the controller derives repo identity by parsing issue IDs (format: `owner-repo-number`), which is fragile for repos with hyphens in the name (e.g., `acme-my-widgets-42` could be `acme/my-widgets` or `acme-my/widgets`). The state machine already preserves canonical `IssueSource` metadata (owner, repo, number, url) through the collect pipeline — this should be used instead.

## Solution

Two changes that together enable multi-repo issue routing:

### 1. Multi-board collection (daemon)

The daemon accepts additional project board IDs via `LEGION_EXTRA_PROJECTS` environment variable. The `/state/fetch-and-collect` endpoint fetches from all configured boards, merges results, and deduplicates by canonical issue identity (`source.owner/source.repo/source.number`). Primary board wins on conflicts.

### 2. Source-based routing (controller + daemon)

- **Controller skill** (primary path): Uses `IssueState.source` (owner, repo) from the collect response for `--repo` in dispatch commands. The controller MUST always pass `--repo` explicitly — it has the freshest data from the collect response.
- **Daemon fallback** (legacy/manual callers): `POST /workers` auto-resolves repo from `issueStateCache` when `--repo` is not provided. This is a convenience for `legion dispatch` CLI usage, NOT the primary controller path.

## Why Not Pattern Matching?

The original issue proposed regex pattern matching on issue IDs mapped to filesystem paths. This is unnecessary because:

1. `IssueState.source` already provides canonical repo metadata (populated by `GitHubTracker.parseIssues()`)
2. The state machine preserves `source` through the entire collect pipeline — it's in the API response
3. `IssueSource.owner` and `IssueSource.repo` are unambiguous — no parsing heuristics needed
4. `ensureWorkspace()` already handles repo clone → workspace creation from `owner/repo` — no filesystem path mapping needed
5. Pattern matching adds configuration complexity for no benefit over canonical metadata

### Multi-board collection
- [ ] Daemon accepts `LEGION_EXTRA_PROJECTS` env var (comma-separated `owner/project-number` values, e.g. `sjawhar/1,other-org/3`). Whitespace is trimmed, empty segments ignored, duplicate entries deduplicated preserving order.
- [ ] `/state/fetch-and-collect` fetches from primary `LEGION_ID` board AND all `LEGION_EXTRA_PROJECTS` boards. Response includes issues from all successfully-fetched boards.
- [ ] Issues appearing on multiple boards are deduplicated by canonical identity (`source.owner/source.repo/source.number`). Primary board's data wins (status, labels, etc.). Precedence order: primary first, then extras in env declaration order.
- [ ] When the same issue appears on multiple boards with different project statuses, a warning is logged: `[fetch-and-collect] duplicate issue {issueId} on boards {board1}, {board2} — using {board1} (primary)`.
- [ ] If one extra board fetch fails, the error is logged and collection continues with remaining boards. The response still succeeds (HTTP 200) with issues from successful boards.
- [ ] If ALL board fetches fail (including primary), the endpoint returns HTTP 500 `fetch_and_collect_failed` (not an empty success — the controller must distinguish "no issues" from "tracker unavailable").
- [ ] Validation at daemon startup: each `LEGION_EXTRA_PROJECTS` entry must match `^[^/]+/\d+$`. Malformed entries cause startup failure with descriptive error.
- [ ] Logging on fetch: `[fetch-and-collect] fetching from N boards: board1, board2, ...`

### Source-based routing
- [ ] Controller dispatch ALWAYS passes `--repo` using `IssueState.source.owner`/`source.repo` from the collect response. This is the primary dispatch path — do not rely on daemon auto-resolve.
- [ ] `POST /workers` auto-resolves repo from `issueStateCache[issueId].source` when `repo` is not provided in request body. This is a fallback for manual CLI callers.
- [ ] Auto-resolve returns HTTP 400 `missing_repo` with message `"provide --repo or ensure issue appears in collected state"` on: cache miss, source is null, or daemon just started (empty cache).
- [ ] Logging on auto-resolve: `[dispatch] auto-resolved repo {owner}/{repo} from issue state for {issueId}`

## Testing Infrastructure

**Available:**
- Bun test runner with comprehensive mocking patterns
- Existing `fetchGitHubProjectItems` mock patterns in server test suite
- `issueStateCache` is in-memory, directly testable
- Controller skill changes are markdown — verified by reading

**Gaps:** None — all changes are testable with existing infrastructure.

### Multi-board collection
- [ ] Unit: `LEGION_EXTRA_PROJECTS` parsing — valid, invalid format, empty string, whitespace, duplicate entries, missing env var
- [ ] Unit: startup validation — malformed entry causes startup failure with descriptive error
- [ ] Unit: fetch-and-collect with 2 boards — issues from each board appear in result
- [ ] Unit: deduplication — same issue (`source.owner/source.repo/source.number`) on two boards, primary board's data used
- [ ] Unit: divergent status — same issue on two boards with different statuses, warning logged, primary wins
- [ ] Unit: graceful degradation — one extra board fetch fails (throws), issues from primary and other extras still returned (HTTP 200)
- [ ] Unit: total failure — primary AND all extra boards fail → HTTP 500 (not empty 200)

### Source-based routing
- [ ] Unit: `POST /workers` auto-resolves repo from cache when `repo` not provided
- [ ] Unit: `POST /workers` returns 400 when `repo` not provided and cache is empty (daemon just started)
- [ ] Unit: `POST /workers` returns 400 when `repo` not provided and `source` is null in cache entry
- [ ] Unit: `POST /workers` returns 400 when `repo` not provided and issue not in cache
- [ ] Unit: verify controller dispatch template in SKILL.md correctly reads `source` from collect response

## Files Affected

- `packages/daemon/src/daemon/config.ts` — add `extraProjects?: string[]` to DaemonConfig, parse from `LEGION_EXTRA_PROJECTS`
- `packages/daemon/src/daemon/server.ts` — multi-board fetch loop in `/state/fetch-and-collect`, auto-resolve in `POST /workers`
- `.opencode/skills/legion-controller/SKILL.md` — dispatch template uses `source` metadata, always passes `--repo`
- Test files for the above

### Multi-board fetch-and-collect (server.ts)

```typescript
// Collect all project board IDs (primary first, then extras in declaration order)
const primaryId = (payload.legionId as string) ?? opts.legionId;
const allProjects = [primaryId, ...(opts.extraProjects ?? [])];
console.log(`[fetch-and-collect] fetching from ${allProjects.length} boards: ${allProjects.join(", ")}`);

// Fetch from each board
const allRawResults: { project: string; items: unknown }[] = [];
const fetchErrors: string[] = [];

for (const project of allProjects) {
  const [owner, numStr] = project.split("/");
  const projectNumber = Number(numStr);
  try {
    const items = await fetchGitHubProjectItems(owner, projectNumber);
    allRawResults.push({ project, items });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[fetch-and-collect] failed to fetch from ${project}: ${msg}`);
    fetchErrors.push(project);
  }
}

// All boards failed → 500 (not empty success)
if (allRawResults.length === 0) {
  return serverError(`fetch_and_collect_failed: all boards failed (${fetchErrors.join(", ")})`);
}

// Parse each board's items, merge with dedup (primary first = wins)
// Dedupe key: source.owner/source.repo/source.number (canonical issue identity)
```

### Auto-resolve repo in POST /workers (server.ts)

```typescript
// Current: repo XOR workspace required
// New: if neither provided, try auto-resolve from cache (fallback for CLI callers)
if (typeof repo !== "string" && typeof workspace !== "string") {
  const cachedState = issueStateCache.get(normalizedIssueId);
  if (cachedState?.source) {
    repo = `${cachedState.source.owner}/${cachedState.source.repo}`;
    console.log(`[dispatch] auto-resolved repo ${repo} from issue state for ${issueId}`);
  } else {
    return badRequest("missing_repo: provide --repo or ensure issue appears in collected state");
  }
}
```

### Controller skill update (SKILL.md)

The controller MUST always pass `--repo` from `IssueState.source` — this is the primary path:

```bash
# Read canonical source from collect response (fresh data, no cache dependency):
SOURCE_OWNER=$(echo "$COLLECTED" | jq -r ".issues.\"$ISSUE_ID\".source.owner // empty")
SOURCE_REPO=$(echo "$COLLECTED" | jq -r ".issues.\"$ISSUE_ID\".source.repo // empty")

# Always pass --repo explicitly:
legion dispatch "$ISSUE_ID" "$MODE" \
  --repo "${SOURCE_OWNER}/${SOURCE_REPO}" \
  --prompt "Invoke the /legion-worker skill for $MODE mode for $ISSUE_ID (github backend, repo: ${SOURCE_OWNER}/${SOURCE_REPO}). Before starting, check for project-specific skills that may be relevant to this work."
```

## Design Decisions

1. **`LEGION_EXTRA_PROJECTS` vs modifying `LEGION_ID`**: Separate env var because `LEGION_ID` is used for path resolution, session IDs, and daemon identity — changing its semantics would break those.
2. **Primary board wins on dedup**: The primary `LEGION_ID` board is the declared system of record for workflow state. Extras provide visibility into other repos' issues but the primary board's status/labels are authoritative. Divergent statuses are logged as warnings.
3. **Controller MUST pass `--repo`** (not rely on auto-resolve): The controller has the freshest data from the collect response. Auto-resolve depends on daemon cache which could be stale after restart. Auto-resolve exists as a convenience for manual `legion dispatch` CLI usage only.
4. **Graceful degradation vs fail-closed**: Individual board failures are graceful (log + continue). ALL boards failing is fail-closed (500, not empty 200) — the controller must distinguish "no issues" from "tracker unavailable."
5. **Dedupe by canonical identity**: `source.owner/source.repo/source.number` is the dedupe key, not project item ID. This correctly identifies the same GitHub issue across multiple project boards.

## Dependencies

- ~~Requires issue #91 (multi-repo dispatch support)~~ — already implemented and closed
- No other blocking dependencies

## Relevant Skills for Downstream Workers

- `github` (GitHub CLI operations)
- `using-jj` (version control)

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #92

## Summary

Adds GitHub multi-project collection and source-based repo routing so the daemon can track issues across multiple project boards, and the controller dispatches workers with canonical repo identity from `IssueState.source`.

### Multi-board collection
- Daemon accepts `LEGION_EXTRA_PROJECTS` env var (comma-separated `owner/project-number` values)
- `/state/fetch-and-collect` fetches from primary + extra boards, deduplicates by canonical source identity (`source.owner/source.repo#source.number`), primary board wins on conflicts
- Graceful degradation: one board fail → 200 with remaining issues, all fail → 500

### Source-based routing  
- Controller skill uses `IssueState.source` from collect response for ALL repo-scoped operations (all 32 `$OWNER/$REPO` references updated to `$ISSUE_REPO`)
- `POST /workers` auto-resolves repo from `issueStateCache` when `--repo` not provided (CLI fallback)
- Returns 400 `missing_repo` on cache miss, null source, or empty cache

### Testing
- 12 new config tests (LEGION_EXTRA_PROJECTS parsing edge cases)
- 9 new server tests (multi-board fetch, dedup, degradation, auto-resolve)
- All 166 tests pass, 0 failures

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 12/12 acceptance criteria verified

### Environment Verification
- `bun test packages/daemon/`: **768 pass, 1 fail** (pre-existing failure in `index.test.ts:1195` — "passes controller env vars to adapter.start on startup" — confirmed unrelated to this PR, documented in implement handoff)
- `tsc --noEmit`: **Clean** (only TS5101 deprecation warning for `baseUrl`, no type errors)
- `biome check`: **Clean** (no fixes needed)

### Results

| # | Criterion | Status | Evidence |
|---|-----------|--------|----------|
| 1 | `LEGION_EXTRA_PROJECTS` env var: comma-separated, whitespace trimmed, empty segments ignored, dedup preserving order | ✅ | `config.ts:42,79-101,137` — `parseExtraProjects()` splits on comma, trims, skips blanks, deduplicates via `Set` preserving insertion order. Pattern: `/^[^\/]+\/\d+$/` |
| 2 | `/state/fetch-and-collect` fetches from primary + all extras | ✅ | `server.ts:1350-1377` — `boardIds = [legionId, ...(opts.extraProjects ?? [])]`, loop fetches each board. `index.ts:472` wires `config.extraProjects` to `startServer()` |
| 3 | Dedup by canonical identity, primary board wins | ✅ | `server.ts:1386-1404` — `seenSources` map keyed by `source.owner/source.repo#source.number` (lowercased). First-seen wins; primary board is always first in `boardIds` |
| 4 | Warning on divergent status across boards | ✅ | `server.ts:1396-1398` — `console.warn("[fetch-and-collect] duplicate issue ... on boards ... — using ... (primary)")`. Test `server.test.ts:1605-1641` verifies warning regex with different-status duplicates. See P3 observation below |
| 5 | One extra board fail → graceful degradation (HTTP 200) | ✅ | `server.ts:1372-1383` — per-board try/catch; error logged, collection continues. Only returns 500 when `collectedBoards.length === 0` |
| 6 | ALL boards fail → HTTP 500 (not empty 200) | ✅ | `server.ts:1379-1383` — `if (collectedBoards.length === 0) return serverError(...)`. Test `server.test.ts:1668-1688` verifies with both boards throwing |
| 7 | Startup validation: malformed entries cause startup failure | ✅ | `config.ts:42,92-94` — `EXTRA_PROJECT_PATTERN.test(entry)` throws on mismatch. `loadConfig()` is called during `startDaemon()`, so throw = startup failure. 12 config tests cover edge cases |
| 8 | Logging on fetch: `[fetch-and-collect] fetching from N boards: ...` | ✅ | `server.ts:1351-1353` — exact format `[fetch-and-collect] fetching from ${boardIds.length} boards: ${boardIds.join(", ")}` |
| 9 | Controller dispatch ALWAYS passes `--repo` using `IssueState.source` | ✅ | `SKILL.md:264-283` — extracts `SOURCE_OWNER`, `SOURCE_REPO`, `ISSUE_REPO` from collect response. All 32 `$OWNER/$REPO` refs replaced with `$ISSUE_REPO`. Dispatch template at `SKILL.md:695-718` uses `--repo "$ISSUE_REPO"` |
| 10 | `POST /workers` auto-resolves repo from cache when not provided | ✅ | `server.ts:644-656` — checks `issueStateCache.get(normalizedIssueId)`, sets `repo` from `cachedState.source.owner/source.repo` |
| 11 | Auto-resolve returns 400 `missing_repo` on cache miss, null source, empty cache | ✅ | `server.ts:652-655` — returns `badRequest("missing_repo: provide --repo or ensure issue appears in collected state")`. Tests at `server.test.ts:469-549` verify 400 for all three cases (empty cache, null source via Linear issue, cache miss) |
| 12 | Logging on auto-resolve | ✅ | `server.ts:648-650` — `console.log("[dispatch] auto-resolved repo ${repo} from issue state for ${issueId}")` |

### Test Quality Critique

**Config tests (12 new in `config.test.ts:190-260`):**
- ✅ **No mocking** — tests `loadConfig()` directly with env var overrides. Clean, focused assertions.
- ✅ **Good edge case coverage** — empty string, single entry, multiple entries, whitespace trimming, empty segments, duplicate dedup, and 5 distinct malformed-entry scenarios.
- ✅ **Verifies behavior, not just types** — asserts both successful parsing AND throws on invalid input.

**Server tests (9 new in `server.test.ts:1522-1689`):**
- ✅ **DI pattern for test seam** — uses injectable `fetchProjectItems` instead of `mock.module()`, which avoids the global mock leakage documented in `docs/solutions/testing/bun-fetch-mocking-patterns.md`. Good choice.
- ✅ **Multi-board coverage** — primary only, primary + extra merged, dedup with primary winning, one fail → 200, all fail → 500.
- ✅ **Dedup warning verified** — uses `console.warn` spy and regex assertion to confirm warning message.

**Auto-resolve tests (4 new in `server.test.ts:407-549`):**
- ✅ **Cache hit with valid source** — populates cache via `/state/collect`, then dispatches without `--repo`. Verifies workspace resolved correctly.
- ✅ **Null source via Linear issue** — populates cache with Linear backend (which has no `source`), verifies 400. Smart approach to exercise real null-source path.
- ✅ **Cache miss** — populates cache with different issue ID, verifies 400 for non-matching issue.
- ✅ **Empty cache** — dispatches immediately without any collect, verifies 400.

**P2: No negative test for duplicate with IDENTICAL statuses.** The dedup test only uses different statuses ("Todo" vs "In Progress"). A test with identical statuses would confirm the warning fires for all duplicates (current behavior) and serve as a regression guard if someone later adds a status-comparison gate.

### Documentation Feedback
- PR description is clear and comprehensive — covers both features, design decisions, and testing approach.
- Code comments explain the DI pattern motivation (mock.module leakage) and the buildCollectedState legionId fix.
- No README or user-facing doc changes needed — these are internal daemon/controller changes.

### Observations
- **P3: Duplicate warning fires for ALL duplicates, not just different-status ones.** The code at `server.ts:1391-1399` warns whenever the same canonical source appears on multiple boards, regardless of whether their project statuses differ. This is stricter than the spec requires (superset behavior) and is harmless, but the warning message says "using {firstBoard} (primary)" even when `firstBoard` is an extra board (not the primary). This is cosmetically misleading if the primary fetch failed.
- **P3: Extra work — titles cache.** `fetch-and-collect` builds and returns a `titles` map and populates `issueTitleCache`. This wasn't in the acceptance criteria but is used by the dashboard endpoint. Harmless.
- **P3: 1 pre-existing test failure** in `daemon/__tests__/index.test.ts:1195` — "passes controller env vars to adapter.start on startup" — documented in implement handoff as unrelated to this PR.

### Config Compliance
- require_specific_task: false — N/A (default, not required)
- require_taiga_evidence: false — N/A (default, not required)

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 3 suggestions

**Verdict: Approved to merge.** No blocking issues found.

---

### Review Scope

Reviewed all 6 changed files against the 16 acceptance criteria in issue #92. Cross-referenced handoff data from architect, plan, implement, and test phases. Verified Oracle analysis of dedup logic correctness and `buildCollectedState` parameter choice.

### Acceptance Criteria Verification

All 16 acceptance criteria are addressed:

**Multi-board collection (8/8):**
- ✅ `LEGION_EXTRA_PROJECTS` env var parsing with validation (`config.ts`)
- ✅ Multi-board fetch in `/state/fetch-and-collect` (`server.ts`)
- ✅ Deduplication by canonical source identity (`source.owner/source.repo#source.number`)
- ✅ Primary board wins on conflicts (first-in-order precedence)
- ✅ Graceful degradation (one board fail → 200, all fail → 500)
- ✅ Startup validation (`^[^/]+/\d+$` pattern)
- ✅ Fetch logging (`[fetch-and-collect] fetching from N boards: ...`)
- ⚠️ Duplicate warning fires for ALL duplicates, not just different-status ones (see P3 #1)

**Source-based routing (4/4):**
- ✅ Controller SKILL.md uses `IssueState.source` for all 32 repo-scoped ops (`$ISSUE_REPO`)
- ✅ `POST /workers` auto-resolves repo from `issueStateCache`
- ✅ Auto-resolve returns 400 on cache miss / null source / empty cache
- ✅ Auto-resolve logging (`[dispatch] auto-resolved repo ...`)

### Key Design Decisions (Verified Correct)
- **`buildCollectedState(issuesData, opts.legionId)`** uses daemon config legionId (not request-resolved). This is correct — `decision.ts` derives deterministic session IDs from this value, and issue #92 explicitly treats `LEGION_ID` as the stable session identity namespace.
- **DI for `fetchProjectItems`** instead of `mock.module()` — justified by Bun's global mock leakage across test files.
- **Pre-existing `$REPO` undefined** in controller SKILL.md — all 32 references correctly updated to `$ISSUE_REPO`.

### P3 Findings

**1. Dedup warning fires unconditionally** (`server.ts:1396`)
AC specifies warnings for "different project statuses"; code warns on ALL duplicates. Superset behavior — functionally harmless but produces extra log noise for same-status duplicates. Consider storing `{boardId, status}` in `seenSources` and only warning when statuses differ.

**2. Warning message says "(primary)" unconditionally** (`server.ts:1397`)
If primary board fetch fails and only extras succeed, the first extra wins on dedup but the message still says "using {board} (primary)". Cosmetically misleading. Consider "using first fetched" or only appending "(primary)" when `existingBoard === legionId`.

**3. No test for same-status duplicate dedup** (`server.test.ts:1605`)
Dedup test only covers different-status case (Todo vs In Progress). Should also test that same-status duplicates are deduped silently (without warning, per the narrowed AC).

### Evidence
- CI status: lint ✓, test ✓, typecheck ✓ — all passing
- No bot review comments to triage
- No stale `$OWNER/$REPO` references remain in controller SKILL.md
- 12 new config tests + 9 new server tests covering edge cases
- Test phase observations (P2 test gap, P3 cosmetic items) align with findings above

File: packages/daemon/src/daemon/server.ts [hunk 1]
```diff
@@ -1324,56 +1337,105 @@ export function startServer(opts: ServerOptions): {
...
+              for (const { boardId, rawIssues } of collectedBoards) {
+                const boardIssues = tracker.parseIssues(rawIssues);
+                for (const issue of boardIssues) {
+                  if (issue.source) {
+                    const sourceKey =
+                      `${issue.source.owner}/${issue.source.repo}#${issue.source.number}`.toLowerCase();
+                    const existingBoard = seenSources.get(sourceKey);
+                    if (existingBoard) {
+                      console.warn(
+                        `[fetch-and-collect] duplicate issue ${sourceKey} on boards ${existingBoard}, ${boardId} — using ${existingBoard} (primary)`
```
Comment (sjawhar) [inline packages/daemon/src/daemon/server.ts:1397]: **[P3]:** Dedup warning fires for ALL duplicates, not just different-status ones. AC specifies "When the same issue appears on multiple boards with **different project statuses**, a warning is logged." Consider storing `{boardId, status}` in `seenSources` and only `console.warn(...)` when the duplicate's status differs from the first copy's.

Comment (sjawhar) [inline packages/daemon/src/daemon/server.ts:1397]: **[P3]:** Warning says `(primary)` unconditionally, but `existingBoard` may be an extra board if the primary fetch failed. Consider neutral wording like `(first fetched)` or only appending `(primary)` when `existingBoard === legionId`.

File: packages/daemon/src/daemon/__tests__/server.test.ts [hunk 2]
```diff
@@ -1365,6 +1518,175 @@ describe("daemon server", () => {
...
+          issues: Record<string, { status: string }>;
+          titles: Record<string, string>;
+        };
+        expect(body.issues["acme-widgets-10"]).toBeDefined();
+        expect(body.issues["other-repo-5"]).toBeDefined();
+        expect(body.titles["acme-widgets-10"]).toBe("Primary issue");
+        expect(body.titles["other-repo-5"]).toBe("Extra issue");
+      });
+
+      it("deduplicates by canonical identity with primary board winning", async () => {
```
Comment (sjawhar) [inline packages/daemon/src/daemon/__tests__/server.test.ts:1605]: **[P3]:** This test covers different-status duplicates (Todo vs In Progress) — good. Missing a companion test for same-status duplicates to verify they are deduped silently (or if the intent is to warn on all duplicates, that should be reflected in the AC).
